cli/command/completions.go

```diff
@@ -0,0 +1,92 @@
+package command
+
+import (
+	"os"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/filters"
+	"github.com/spf13/cobra"
+)
+
+// ShowContainerIDs if true will offer completion on container IDs, not just names
+var ShowContainerIDs = os.Getenv("DOCKER_COMPLETION_SHOW_CONTAINER_IDS") == "yes"
+
+// CompletionFn a function to be used by cobra command to offer command line completion
+type CompletionFn func(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective)
+
+// ContainerNamesCompletionFn offers completion for container names and IDs
+// By default, only names are returned.
+// Set DOCKER_COMPLETION_SHOW_CONTAINER_IDS=yes to also complete IDs.
+func ContainerNamesCompletionFn(dockerCli Cli, all bool, filters ...func(types.Container) bool) CompletionFn {
+	return func(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
```

Comment:
I haven't tested so apologies if I'm not reading the code right.
I believe this function will complete container names, no matter what the `args` field is.
This would mean that doing
```
docker exec myContainer <TAB>
```
would continue suggesting other container names.
You may want to have a `repeat bool` argument for this function which would indicate if container names should continue to be suggested, or if they should only be suggested for the first argument.  And then base your logic on `len(args)`.

This may apply to other completion function as well.